### PR TITLE
Spell/Script: Scourge Disguise and Haunting Phantoms

### PR DESCRIPTION
**Pull request by PolarCookie**
**Changes proposed:**
Found by lucky concidence https://github.com/TrinityCore/TrinityCore/blob/master/src/server/scripts/EasternKingdoms/ShadowfangKeep/shadowfang_keep.cpp#L324-L361.
Used it as a base to solve https://github.com/TrinityCore/TrinityCore/issues/22687 https://github.com/TrinityCore/TrinityCore/issues/22662

**Target branch(es):**
3.3.5

**Issues addressed:**
https://github.com/TrinityCore/TrinityCore/issues/22687 can be closed.
https://github.com/TrinityCore/TrinityCore/issues/22662 this PR does not make the NPCs actually cast this spell, just fixes the spell script itself.

**Tests performed:**
Builds and works in-game.

**Known issues and TODO list:**
- [x] When Scourge Disguise Expiring is triggered, a warning should flash across the screen; "Scourge Disguise Failing! Find a safe place!".
https://www.wowhead.com/item=38699/ensorcelled-choker#comments:id=458778

### CODE REVIEW

File: sql/updates/world/3.3.5/9999_99_99_99_world.sql [hunk 1]
```diff
@@ -0,0 +1,10 @@
+DELETE FROM `spell_script_names` WHERE `spell_id` IN (16336, 51971, 52010);
+INSERT INTO `spell_script_names` (`spell_id`,`ScriptName`) VALUES
+(16336, 'spell_stratholme_haunting_phantoms'),
+(51971, 'spell_scourge_disguise_instability'),
+(52010, 'spell_scourge_disguise_expiring');
+
+DELETE FROM `spell_linked_spell` WHERE `spell_trigger` = 51966;
+INSERT INTO `spell_linked_spell` (`spell_trigger`, `spell_effect`, `type`, `comment`) VALUES
+(51966, 51971, 1, 'Scourge Disguise applies Scourge Disguise Instability');
```
Comment (Jildor) [inline sql/updates/world/3.3.5/9999_99_99_99_world.sql]: I think this is considered hacky and will not be accepted 😄

Comment (Keader) [reply sql/updates/world/3.3.5/9999_99_99_99_world.sql]: this table should be dropped... :)

Comment (Keader) [reply sql/updates/world/3.3.5/9999_99_99_99_world.sql]: In cases that is: Spell 1 -> Trigger Spell 2, it dont even check if Spell1 realy hit target ._.

Comment (Warpten) [reply sql/updates/world/3.3.5/9999_99_99_99_world.sql]: Yeah this table blows

Comment (PolarCookie) [reply sql/updates/world/3.3.5/9999_99_99_99_world.sql]: I'm not gonna complain that it hasn't been nuked already in that case, but could you at least make a note of it on https://trinitycore.atlassian.net/wiki/spaces/tc/pages/2130132/spell+linked+spell?

File: sql/updates/world/3.3.5/9999_99_99_99_world.sql [hunk 2]
```diff
@@ -0,0 +1,11 @@
+DELETE FROM `spell_script_names` WHERE `spell_id` IN (16336, 51966, 51971, 52010, 54089);
+INSERT INTO `spell_script_names` (`spell_id`,`ScriptName`) VALUES
+(16336, 'spell_stratholme_haunting_phantoms'),
+(51966, 'spell_scourge_disguise'),
+(51971, 'spell_scourge_disguise_instability'),
+(52010, 'spell_scourge_disguise_expiring'),
+(54089, 'spell_drop_disguise');
+
+
+DELETE FROM `spell_linked_spell` WHERE `spell_trigger` = 54089;
```
Comment (illfated) [inline sql/updates/world/3.3.5/9999_99_99_99_world.sql]: suggestion: move this line on top and delete at least 2, maybe all 3 blank lines.

File: src/server/scripts/EasternKingdoms/Stratholme/stratholme.cpp [hunk 3]
```diff
@@ -332,10 +333,49 @@ class spell_ysida_saved_credit : public SpellScript
...
+
+    void Register() override
+    {
+        DoEffectCalcPeriodic += AuraEffectCalcPeriodicFn(spell_stratholme_haunting_phantoms::CalcPeriodic, EFFECT_0, SPELL_AURA_DUMMY);
+        OnEffectPeriodic += AuraEffectPeriodicFn(spell_stratholme_haunting_phantoms::HandleDummyTick, EFFECT_0, SPELL_AURA_DUMMY);
+        OnEffectUpdatePeriodic += AuraEffectUpdatePeriodicFn(spell_stratholme_haunting_phantoms::HandleUpdatePeriodic, EFFECT_0, SPELL_AURA_DUMMY);
+    }
+};
+
+
```
Comment (illfated) [inline src/server/scripts/EasternKingdoms/Stratholme/stratholme.cpp]: Cosmetic/codestyle: redundant blank line.

**Comment by PolarCookie** [comment]
I'm looking at sniffs for the instability timer, found https://db.rising-gods.de/?spell=52335 being used by https://db.rising-gods.de/?npc=28778, related to https://db.rising-gods.de/?quest=12676.
Putting it out there, but not going to fix here.

**Comment by PolarCookie** [comment]
I'm happy with this as is.

File: src/server/scripts/Northrend/zone_zuldrak.cpp [hunk 4]
```diff
@@ -949,6 +949,100 @@ class npc_storm_cloud : public CreatureScript
...
+
+    void Register() override
+    {
+        DoEffectCalcPeriodic += AuraEffectCalcPeriodicFn(spell_scourge_disguise_instability::CalcPeriodic, EFFECT_0, SPELL_AURA_DUMMY);
+        OnEffectPeriodic += AuraEffectPeriodicFn(spell_scourge_disguise_instability::HandleDummyTick, EFFECT_0, SPELL_AURA_DUMMY);
+        OnEffectUpdatePeriodic += AuraEffectUpdatePeriodicFn(spell_scourge_disguise_instability::HandleUpdatePeriodic, EFFECT_0, SPELL_AURA_DUMMY);
+    }
+};
+
+static std::string scourge_disguise_expiring_warning = "Disguise Failing! Avoid Scourge Contact!";
```
Comment (ccrs) [inline src/server/scripts/Northrend/zone_zuldrak.cpp]: ugh

File: src/server/scripts/EasternKingdoms/Stratholme/stratholme.cpp [hunk 5]
```diff
@@ -332,10 +333,48 @@ class spell_ysida_saved_credit : public SpellScript
...
+    void CalcPeriodic(AuraEffect const* /*aurEff*/, bool& isPeriodic, int32& amplitude)
+    {
+        isPeriodic = true;
+        amplitude = irand(30, 90) * IN_MILLISECONDS;
+    }
+
+    void HandleDummyTick(AuraEffect const* /*aurEff*/)
+    {
+        if (irand(0, 1))
```
Comment (ccrs) [inline src/server/scripts/EasternKingdoms/Stratholme/stratholme.cpp]: there is a rand method that returns bool, cant remember hows it called though

File: src/server/scripts/Northrend/zone_zuldrak.cpp (hunk 4 above)
Comment (illfated) [reply src/server/scripts/Northrend/zone_zuldrak.cpp]: Yeah. Text belongs in DB (preferred / if possible) or in some kind of shared define list.

Comment (Aokromes) [reply src/server/scripts/Northrend/zone_zuldrak.cpp]: broadcast_text ID 28891

Comment (PolarCookie) [reply src/server/scripts/Northrend/zone_zuldrak.cpp]: thanks, @Aokromes!
That's how you do feedback right!

File: src/server/scripts/EasternKingdoms/Stratholme/stratholme.cpp (hunk 5 above)
Comment (PolarCookie) [reply src/server/scripts/EasternKingdoms/Stratholme/stratholme.cpp]: I didn't find it.

File: src/server/scripts/Northrend/zone_zuldrak.cpp [hunk 6]
```diff
@@ -1017,7 +1017,7 @@ class spell_scourge_disguise_expiring : public AuraScript
     {
         if (Player* player = GetTarget()->ToPlayer())
             if (Aura* aura = player->GetAura(SPELL_SCOURGE_DISGUISE))
-                player->Talk(scourge_disguise_expiring_warning, CHAT_MSG_RAID_BOSS_WHISPER, LANG_UNIVERSAL, 0, nullptr);
+                player->Talk(TEXT_DISGUISE_WARNING, CHAT_MSG_RAID_BOSS_WHISPER, 0, nullptr);
```
Comment (Aokromes) [inline src/server/scripts/Northrend/zone_zuldrak.cpp]: give a look to https://github.com/TrinityCore/TrinityCore/commit/5f541491074765a346c7d0a83edb5f1fbf9240b1 maybe it's better way.

File: src/server/scripts/EasternKingdoms/Stratholme/stratholme.cpp (hunk 5 above)
Comment (P-Kito) [reply src/server/scripts/EasternKingdoms/Stratholme/stratholme.cpp]: `roll_chance_i(50)` ?

File: src/server/scripts/Northrend/zone_zuldrak.cpp (hunk 6 above)
Comment (PolarCookie) [reply src/server/scripts/Northrend/zone_zuldrak.cpp]: I didn't become any wiser by reading that, `Unit::Talk` and `Battleground::SendBroadcastText` do roughly the same except for how they pick targets.

Comment (P-Kito) [reply src/server/scripts/Northrend/zone_zuldrak.cpp]: I am not sure how your link should help him @Aokromes

File: src/server/scripts/EasternKingdoms/Stratholme/stratholme.cpp (hunk 5 above)
Comment (P-Kito) [inline src/server/scripts/EasternKingdoms/Stratholme/stratholme.cpp]: `roll_chance_i(50)` ? Is this what you meant @ccrs

Comment (illfated) [reply src/server/scripts/EasternKingdoms/Stratholme/stratholme.cpp]: Thanks, I was wondering if ccrs meant a TC function or general/external.

Comment (ccrs) [reply src/server/scripts/EasternKingdoms/Stratholme/stratholme.cpp]: yeah

File: src/server/scripts/Northrend/zone_zuldrak.cpp (hunk 6 above)
Comment (illfated) [reply src/server/scripts/Northrend/zone_zuldrak.cpp]: IMHO those battleground broadcast examples should be better suited for targeting multiple players.
Would it be helpful to look at existing code where `CHAT_MSG_RAID_BOSS_WHISPER` is used already?
https://github.com/TrinityCore/TrinityCore/blob/f1a1acdc89a6521fddca98cc8ef8d6e7ed0b9495/src/server/game/Texts/CreatureTextMgr.cpp#L324-L335

**Comment by Aokromes** [comment]
there is something missing here?

**Comment by PolarCookie** [comment]
not that I can remember.

**Pull request merged**